qiskit/pulse/instructions/call.py

```diff
@@ -38,35 +38,87 @@ def __init__(self, subroutine,
         .. note:: Inline subroutine is mutable. This requires special care for modification.
 
         Args:
-            subroutine (Schedule): A program subroutine to be referred to.
+            subroutine (Union[Schedule, ScheduleBlock]): A program subroutine to be referred to.
             value_dict: Mapping of parameter object to assigned value.
             name: Unique ID of this subroutine. If not provided, this is generated based on
                 the subroutine name.
-        """
-        if name is None:
-            name = f"{self.prefix}_{subroutine.name}"
 
-        super().__init__((subroutine,), None,
-                         channels=tuple(subroutine.channels),
-                         name=name)
-        if value_dict:
-            self.assign_parameters(value_dict)
+        Raises:
+            PulseError: If subroutine is not valid data format.
+        """
+        from qiskit.pulse.schedule import ScheduleBlock, Schedule
+
+        if not isinstance(subroutine, (ScheduleBlock, Schedule)):
+            raise PulseError(f'Subroutine {subroutine} is not acceptable data format.')
+
+        # initialize parameter template
+        # TODO remove self._parameter_table
+        self._arguments = dict()
+
+        if subroutine.is_parameterized():
+            for param_expr in subroutine.parameters:
+                if value_dict:
+                    param_assigned = copy.copy(param_expr)
+                    for param in param_expr.parameters:
+                        if param in value_dict:
+                            param_assigned = param_assigned.assign(param, value_dict[param])
+                    self._arguments[param_expr] = param_assigned
+                else:
+                    self._arguments[param_expr] = param_expr
+
+        # create cache data of parameter-assigned subroutine
+        assigned_subroutine = subroutine.assign_parameters(
+            value_dict=self.arguments,
+            inplace=False
+        )
+        self._assigned_cache = tuple((self._get_arg_hash(), assigned_subroutine))
+
+        super().__init__(operands=(subroutine, ), name=name or f"{self.prefix}_{subroutine.name}")
 
     @property
     def duration(self) -> Union[int, ParameterExpression]:
         """Duration of this instruction."""
         return self.subroutine.duration
```

Comment:
`self.subroutine` may be used in the future transfer layer, or in the experiment. Since the subroutine is kept parametrized, we can compactly represent various sub schedules with different assigned parameters.

On the other hand, `self.assigned_subroutine` is required for equivalence check. For example,
```python
amp = Parameter('amp')

with build() as test_sched:
    call(Schedule(Gaussian(160, amp, 40), DriveChannel(0)))

with build() as ref_sched:
    call(Schedule(Gaussian(160, 1.0, 40), DriveChannel(0)))

test_sched.assign_parameters({amp: 0.1}) == ref_sched # False
```
here we expect assigned `test_sched` and `ref_sched` are equal, but they are actually evaluated as different subroutines since one subroutine is still parametrized and other has no parameter. Perhaps `assigned_subroutine` should be a private method.